src/server/command/worldedit/blueprint.zig

```diff
@@ -0,0 +1,170 @@
+	}
+};
+
+pub fn execute(args: []const u8, source: *User) void {
+	var argsList = List([]const u8).init(main.stackAllocator);
+	defer argsList.deinit();
+
+	var splitIterator = std.mem.splitScalar(u8, args, ' ');
+	while(splitIterator.next()) |a| {
+		argsList.append(a);
+	}
+
+	if(argsList.items.len < 1) {
+		source.sendMessage("#ff0000Not enough arguments for /blueprint, expected at least 1.", .{});
+		return;
+	}
+	const subcommand = BlueprintSubCommand.fromString(argsList.items[0]);
+	_ = switch(subcommand) {
+		.save => blueprintSave(argsList.items, source),
+		.delete => blueprintDelete(argsList.items, source),
+		.load => blueprintLoad(argsList.items, source),
+		.list => blueprintList(source),
+		.other => {
+			source.sendMessage("#ff0000Unrecognized subcommand for /blueprint: '{s}'", .{argsList.items[0]});
+		},
+		.empty => {
+			source.sendMessage("#ff0000Missing subcommand for **/blueprint**, usage: {s} ", .{usage});
+		},
+	} catch |err| {
+		std.log.info("Error: {s}", .{@errorName(err)});
+		source.sendMessage("#ff0000Error: {s}", .{@errorName(err)});
+	};
+}
+
+fn blueprintSave(args: []const []const u8, source: *User) !void {
+	if(args.len < 2) {
+		source.sendMessage("#ff0000**/blueprint save** requires FILENAME argument.", .{});
+		return;
+	}
+	if(args.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint save**. Expected 1 argument, FILENAME.", .{});
+		return;
+	}
+
+	if(source.commandData.clipboard) |clipboard| {
+		const storedBlueprint = clipboard.store(main.stackAllocator);
+		defer main.stackAllocator.free(storedBlueprint);
+
+		const fileName: []const u8 = ensureBlueprintExtension(main.stackAllocator, args[1]);
+		defer main.stackAllocator.free(fileName);
+
+		var blueprintsDir = try openDir("blueprints");
+		defer blueprintsDir.close();
+
+		try blueprintsDir.write(fileName, storedBlueprint);
+
+		std.log.info("Saved clipboard to blueprint file: {s}", .{fileName});
+		source.sendMessage("#00ff00Saved clipboard to blueprint file: {s}", .{fileName});
+	} else {
+		source.sendMessage("#ff0000Error: No clipboard content to save.", .{});
+	}
+}
+
+fn ensureBlueprintExtension(allocator: NeverFailingAllocator, fileName: []const u8) []const u8 {
+	if(!std.ascii.endsWithIgnoreCase(fileName, ".blp")) {
+		return std.fmt.allocPrint(allocator.allocator, "{s}.blp", .{fileName}) catch unreachable;
+	} else {
+		return allocator.dupe(u8, fileName);
+	}
+}
+
+fn blueprintDelete(args: []const []const u8, source: *User) !void {
+	if(args.len < 2) {
+		source.sendMessage("#ff0000**/blueprint delete** requires FILENAME argument.", .{});
+		return;
+	}
+	if(args.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint delete**. Expected 1 argument, FILENAME.", .{});
+		return;
+	}
+
+	const fileName: []const u8 = ensureBlueprintExtension(main.stackAllocator, args[1]);
+	defer main.stackAllocator.free(fileName);
+
+	var blueprintsDir = try openDir("blueprints");
+	defer blueprintsDir.close();
```

Comment:
I keep seeing the same thing three times here. I wonder if it would be possible to turn it into a helper function?
Probably not, but maybe you have an idea how it could be done?

Reply:
I am not sure which parts you mean. For `fileName` and `blueprintDir` alone I don't see benefit in extracting them to separate function, it would be almost the same amount of code, you still need to create both, return both and clean up both. As for parameter list checking, this is coincidental, those functions can in the future accept different flags. Ideally this would be abstracted by command parsing library, maybe a home grown one so it makes good use of source.sendMessage.